### PR TITLE
[incubator-kie-drools-5918] [new-parser] Annotations can appear in lh…

### PR DESCRIPTION
**Pull request by tkobayas**
…sOr and lhsAnd

**Issue**:
* https://github.com/apache/incubator-kie-drools/issues/5918

### CODE REVIEW

File: drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/DRLVisitorImpl.java [hunk 1]
```diff
@@ -1032,7 +1038,13 @@ private BaseDescr createAndDescr(ParserRuleContext ctx) {
...
             // So, we need to flatten it so that AndDescr has A(), B() and C() as children.
             List<BaseDescr> flattenedDescrs = flattenAndDescr(descrList);
-            flattenedDescrs.forEach(andDescr::addDescr);
+            flattenedDescrs.forEach(descr -> {
+                if (descr instanceof AnnotationDescr annotationDescr) {
+                    andDescr.addAnnotation(annotationDescr);
+                } else {
+                    andDescr.addDescr(descr);
+                }
+            });
```
Comment (tkobayas) [inline drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/DRLVisitorImpl.java:1048]: I know there is a duplication in this part and the above one (`visitLhsOr`). But `addDescr` method is not defined in a common parent class of `AndDescr` and `OrDescr`, so I cannot refactor them to a reusable clean private method. (Changing Descr class hierarchy sounds also too much).

**Comment by tkobayas** [comment]
@yurloc @mariofusco @gitgabrio Please review, thanks!

**Review by mariofusco — APPROVED** [review]

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java [hunk 2]
```diff
@@ -3990,6 +3990,50 @@ void lhsPatternAnnotation() {
         assertThat(annotationDescr.getSingleValueAsString()).isEqualTo("!*, age");
     }
 
+    @Test
+    void andDescrAnnotation() {
+        ParserTestUtils.enableOldParser();
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java]: Committed unintentionally, I believe.

**Review by yurloc — CHANGES_REQUESTED** [review]
To be compliant with the old parser, we also need to process annotations from each expression joined by infix OR/AND:
```
rule R
    when
       ( Double() 
         and @Annot1 String() 
         and @Annot2 Integer() ) 
    then
end
```

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java (hunk 2 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java]: Thank you! Fixed.

**Comment by tkobayas** [comment]
> To be compliant with the old parser, we also need to process annotations from each expression joined by infix OR/AND:

Thank you! I added test cases and also fixed the implementation to collect the annotations.

**Review by yurloc — APPROVED** [review]
Perfect! Let's merge.

**Pull request merged**